{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/kosarHNajafi/R.Codes/blob/main/NCA.MP.TF.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# Network Component Analysis (NCA)\n",
        "[RTN Bioconductor Tutorial Page](https://bioconductor.org/packages/devel/bioc/vignettes/RTN/inst/doc/RTN.html#ref-Margolin2006b)\n"
      ],
      "metadata": {
        "id": "OdM5T5kB4I2n"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#---Load libraries-----------------\n",
        "library(biomaRt)\n",
        "library(RTN)\n",
        "\n",
        "set.seed(123)\n",
        "#---1.Load Gene Expression Data of 90 metabolic pathways and TFs = a mtrix for counts/assays---------------------------------\n",
        "#load txt. Format\n",
        "DISC.MP.TF <- read.delim(\"~/NCA.ER/DISC.MP.TF.Curated.txt\")\n",
        "View(DISC.MP.TF)\n",
        "\n",
        "#Preparing disc.mp.tf\n",
        "disc.mp.tf <- DISC.MP.TF[,-1]\n",
        "rownames(disc.mp.tf) <- DISC.MP.TF[,1]\n",
        "View(disc.mp.tf)\n",
        "# Replace dots with underscores in column names if needed\n",
        "colnames(disc.mp.tf) <- gsub(\"\\\\.\", \"_\", colnames(disc.mp.tf))\n",
        "\n",
        "#Sort data by columnnames(Sample IDs) and rownames(Gene IDs)\n",
        "disc.mp.tf <- disc.mp.tf[,order(colnames(disc.mp.tf))]\n",
        "disc.mp.tf <- disc.mp.tf[order(rownames(disc.mp.tf)),]\n",
        "View(disc.mp.tf)\n",
        "head(disc.mp.tf)\n",
        "str(disc.mp.tf)\n",
        "dim(disc.mp.tf)\n",
        "print(\"Step 1 completed: Loading Metabolic and TF Gene expression in Discovery set\")\n",
        "\n",
        "#---2.Available METABRIC Clinical File as sampleAnnotation.disc =colData----------------------\n",
        "Metabric_Manual_disc <- read.delim(\"~/NCA/data/METABRIC Clinical.txt\", row.names = 1)\n",
        "View(Metabric_Manual_disc)\n",
        "\n",
        "Metabric_Manual_disc <- Metabric_Manual_disc[order(rownames(Metabric_Manual_disc)),]\n",
        "View(Metabric_Manual_disc)\n",
        "\n",
        "# Replace dots with underscores in column names if needed\n",
        "rownames(Metabric_Manual_disc) <- gsub(\"\\\\-\", \"_\", rownames(Metabric_Manual_disc))\n",
        "View(Metabric_Manual_disc)\n",
        "\n",
        "# Load required library\n",
        "library(dplyr)\n",
        "\n",
        "# Assuming your dataset is named Metabric_Manual_disc\n",
        "# Create a new binary dataframe based on transformations\n",
        "Metabric_Manual_disc <- Metabric_Manual_disc %>%\n",
        "  transmute(\n",
        "    IDs = rownames(Metabric_Manual_disc),\n",
        "    Cohort = Cohort,\n",
        "\n",
        "    OS.time = Overall.Survival..Months.,\n",
        "    OS.event = ifelse(Overall.Survival.Status == \"1:DECEASED\", 1, 0),\n",
        "    DSS.event = ifelse(Patient.s.Vital.Status == \"Died of Disease\", 1, 0),\n",
        "    RFS.event = ifelse(Relapse.Free.Status == \"1:Recurred\", 1, 0),\n",
        "    RFS.time = Relapse.Free.Status..Months.,\n",
        "\n",
        "    Grade = Neoplasm.Histologic.Grade,\n",
        "    Size = Tumor.Size,\n",
        "    LN = Lymph.nodes.examined.positive,\n",
        "    Age = Age.at.Diagnosis,\n",
        "    LN = Lymph.nodes.examined.positive,\n",
        "    Age = Age.at.Diagnosis,\n",
        "\n",
        "\n",
        "    # Subtypes for LumA, LumB, Basal, Her2, Normal based on Pam50 subtype\n",
        "    LumA = ifelse(Pam50...Claudin.low.subtype == \"LumA\", 1, 0),\n",
        "    LumB = ifelse(Pam50...Claudin.low.subtype == \"LumB\", 1, 0),\n",
        "    Basal = ifelse(Pam50...Claudin.low.subtype == \"Basal\", 1, 0),\n",
        "    Her2 = ifelse(Pam50...Claudin.low.subtype == \"Her2\", 1, 0),\n",
        "    Normal = ifelse(Pam50...Claudin.low.subtype == \"Normal\", 1, 0),\n",
        "\n",
        "    # ER and PR status (positive and negative)\n",
        "    `ER+` = ifelse(ER.Status == \"Positive\", 1, 0),\n",
        "    `ER-` = ifelse(ER.Status == \"Negative\", 1, 0),\n",
        "\n",
        "\n",
        "    # Histologic Grade categories G1, G2, G3\n",
        "    G1 = ifelse(Neoplasm.Histologic.Grade == 1, 1, 0),\n",
        "    G2 = ifelse(Neoplasm.Histologic.Grade == 2, 1, 0),\n",
        "    G3 = ifelse(Neoplasm.Histologic.Grade == 3, 1, 0),\n",
        "\n",
        "    # Hormone Therapy (HT)\n",
        "    HT = ifelse(Hormone.Therapy == \"YES\", 1, 0),\n",
        "\n",
        "  )\n",
        "View(Metabric_Manual_disc)\n",
        "\n",
        "# Print head of the colAnnotation for verification\n",
        "head(Metabric_Manual_disc)\n",
        "\n",
        "print(\"Step 2 completed: Manual METABRIC Clinical data\")\n",
        "\n",
        "#---3.disc_sample_mp.tf_ids Sample/Column Annotation-----------------------------\n",
        "disc_sample_mp.tf_ids <- colnames(disc.mp.tf)\n",
        "View(disc_sample_mp.tf_ids)\n",
        "length(disc_sample_mp.tf_ids)\n",
        "\n",
        "sort(disc_sample_mp.tf_ids)\n",
        "View(disc_sample_mp.tf_ids)\n",
        "length(disc_sample_mp.tf_ids)\n",
        "\n",
        "all.equal(disc_sample_mp.tf_ids,colnames(disc.mp.tf))\n",
        "\n",
        "# Find the common sample IDs between the two datasets\n",
        "common_samples_disc <- intersect(rownames(Metabric_Manual_disc), disc_sample_mp.tf_ids)\n",
        "common_samples_disc <- sort(common_samples_disc)\n",
        "View(common_samples_disc)\n",
        "length(common_samples_disc)\n",
        "\n",
        "# Subset and reorder both datasets to only include common samples\n",
        "counts.mp.tf.disc <- disc.mp.tf[, common_samples_disc]\n",
        "all.equal(rownames(disc.mp.tf), rownames(counts.mp.tf.disc))\n",
        "View(counts.mp.tf.disc)\n",
        "dim(counts.mp.tf.disc)\n",
        "\n",
        "sampleAnnotation.disc <- Metabric_Manual_disc[common_samples_disc, ]\n",
        "dim(sampleAnnotation.disc)\n",
        "\n",
        "# Verify that they are aligned\n",
        "all.equal(colnames(counts.mp.tf.disc), rownames(sampleAnnotation.disc))  # Should return TRUE\n",
        "\n",
        "# If the first is not TRUE, you can match up the samples/columns in\n",
        "# counts with the samples/rows in sampleAnnotation.disc like this (which is fine\n",
        "# to run even if the first was TRUE):\n",
        "\n",
        "#tempindex <- match(colnames(counts.mp.tf.disc), rownames(sample_metadata))\n",
        "#sampleAnnotation.disc <- sample_metadata[tempindex, ]\n",
        "\n",
        "#Check again\n",
        "\n",
        "all.equal(colnames(counts.mp.tf.disc), rownames(sampleAnnotation.disc))\n",
        "\n",
        "print(\"Step 3 completed: Discovery SampleAnnotation Prepared\")\n",
        "\n",
        "#---4.Load Ensemble Gene Annotation----------------------------------------------\n",
        "\n",
        "#SHOULD I KEEP OR REMOVE?\n",
        "#1.Accessing the data available in Ensembl by biomaRT\n",
        "browseVignettes(\"biomaRt\")\n",
        "\n",
        "#2.Selecting an Ensembl BioMart database and dataset\n",
        "\n",
        "# Connect to Ensembl database, and query human genes\n",
        "ensembl <- useEnsembl(biomart = \"genes\", dataset = \"hsapiens_gene_ensembl\")\n",
        "\n",
        "\n",
        "#2.1##Step1: Identifying the database you need\n",
        "\n",
        "#find the names of the BioMart services Ensembl is currently providing\n",
        "##and Can be used to connect to the desired BioMart database:\n",
        "listEnsembl()\n",
        "#biomart argument should be given a valid name from the output of listEnsembl()\n",
        "ensembl <- useEnsembl(biomart = \"genes\")\n",
        "\n",
        "\n",
        "#2.2##Step 2: Choosing a dataset\n",
        "\n",
        "#Within the Ensembl dataset each species is a different dataset.\n",
        "#look at which datasets are available in the selected BioMart\n",
        "#by using the function listDatasets()\n",
        "datasets <- listDatasets(ensembl)\n",
        "View(datasets)\n",
        "\n",
        "##Because the listDatasets are so long\n",
        "##in ensembl mart find anything that contain the term ‘hsapiens’\n",
        "\n",
        "searchDatasets(mart = ensembl, pattern = \"hsapiens\")\n",
        "\n",
        "#If you've been through these all before;select a both the database and dataset in one step\n",
        "ensembl <- useDataset(dataset = \"hsapiens_gene_ensembl\", mart = ensembl)\n",
        "\n",
        "\n",
        "#2.3##Ensembl mirror sites\n",
        "#listMarts() to find the biomart\n",
        "ensembl <- useEnsembl(biomart = \"ENSEMBL_MART_ENSEMBL\",\n",
        "                      dataset = \"hsapiens_gene_ensembl\",\n",
        "                      mirror = \"useast\")\n",
        "#UP TO HERE SHALL I KEEP OR CAN I REMOVE IT???\n",
        "#maintain consistent annotation throughout the duration of a project.\n",
        "#2.4##Using archived versions of Ensembl, with no arguments\n",
        "listEnsemblArchives()\n",
        "listEnsembl(version = 113)\n",
        "searchDatasets(mart = ensembl_113, pattern = \"hsapiens\")\n",
        "#to ensure that script you write now will return exactly the same results in the future;\n",
        "##copy the URL from listEnsemblArchive() or www.ensembl.org in host Argument\n",
        "ensembl_113 <- useEnsembl(biomart = \"genes\",\n",
        "                          dataset = \"hsapiens_gene_ensembl\",\n",
        "                          host = \"https://oct2024.archive.ensembl.org\",\n",
        "                          version = 113)\n",
        "\n",
        "print(\"Step 8 completed: Ensembl Gene Information loaded\")\n",
        "\n",
        "#---5.disc_mp.tf Gene Annotation/build a biomaRt query ---------------------------\n",
        "# Extract gene annotations for the corresponding genes from `ensembl_113`\n",
        "filters = listFilters(ensembl_113)\n",
        "View(filters)\n",
        "\n",
        "View(listAttributes(ensembl_113))\n",
        "\n",
        "gene_annot_mp.tf_disc <- getBM(\n",
        "  attributes = c(\"ensembl_gene_id\", \"chromosome_name\", \"start_position\", \"end_position\", \"strand\", \"external_gene_name\",\"description\"),\n",
        "  filters = \"external_gene_name\",\n",
        "  values = rownames(disc.mp.tf),\n",
        "  mart = ensembl_113\n",
        ")\n",
        "\n",
        "#WITH ENTREZ 3185 AND WITHOUT THAT 3152, DON'T MENTION \"entrezgene_id\".\n",
        "View(gene_annot_mp.tf_disc)\n",
        "\n",
        "# Check for duplicated external_gene_names\n",
        "table(duplicated(gene_annot_mp.tf_disc$external_gene_name))\n",
        "\n",
        "write.table(gene_annot_mp.tf_disc,file = file.path(\"~/NCA.ER/\",\"Gene_annot_mp.tf.disc.txt\"),sep = \"\\t\",row.names = FALSE)\n",
        "\n",
        "#Remove Duplicates Manually\n",
        "gene_annot_mp.tf_disc <- read.delim(\"~/NCA.ER/NCA.MP.TF.DISC/GeneAnnotationCurated_Ensembl113.Used.txt\")\n",
        "\n",
        "dim(gene_annot_mp.tf_disc)\n",
        "length(gene_annot_mp.tf_disc)\n",
        "View(gene_annot_mp.tf_disc)\n",
        "dim(counts.mp.tf.disc)\n",
        "\n",
        "#Only Validated IDs Taken\n",
        "\n",
        "# Extract gene IDs from both datasets\n",
        "ensemble_ids <- gene_annot_mp.tf_disc$external_gene_name\n",
        "View(ensemble_ids)\n",
        "\n",
        "ensemble_ids <- sort(ensemble_ids)\n",
        "length(ensemble_ids)\n",
        "\n",
        "disc_mp.tf_gene_ids <- rownames(counts.mp.tf.disc)\n",
        "View(disc_mp.tf_gene_ids)\n",
        "\n",
        "common_genes <- intersect(ensemble_ids, disc_mp.tf_gene_ids) # Should match or be close to 1420\n",
        "common_genes <- sort(common_genes)\n",
        "all.equal(ensemble_ids, common_genes)\n",
        "length(common_genes)\n",
        "\n",
        "# Subset and reorder both datasets to only include common samples\n",
        "counts.mp.tf.disc <- counts.mp.tf.disc[common_genes, ]\n",
        "View(counts.mp.tf.disc)\n",
        "\n",
        "all.equal(ensemble_ids,rownames(counts.mp.tf.disc))\n",
        "\n",
        "#Set rownames for gene annotation, %n% doesn't let your data go NA\n",
        "gene_annot_mp.tf_disc <- gene_annot_mp.tf_disc[gene_annot_mp.tf_disc$external_gene_name %in% common_genes, ]\n",
        "rownames(gene_annot_mp.tf_disc) <- gene_annot_mp.tf_disc$external_gene_name\n",
        "View(gene_annot_mp.tf_disc)\n",
        "all.equal(gene_annot_mp.tf_disc$external_gene_name, rownames(counts.mp.tf.disc))\n",
        "all.equal(gene_annot_mp.tf_disc$external_gene_name,rownames(gene_annot_mp.tf_disc))\n",
        "all.equal(gene_annot_mp.tf_disc$external_gene_name,common_genes)\n",
        "all.equal(common_genes,ensemble_ids)\n",
        "\n",
        "#Sort counts.mp.tf.disc and gene_annot_mp.tf_disc based on rownames which is gene names\n",
        "counts.mp.tf.disc <- counts.mp.tf.disc[order(rownames(counts.mp.tf.disc)), ]\n",
        "gene_annot_mp.tf_disc <- gene_annot_mp.tf_disc[order(rownames(gene_annot_mp.tf_disc)), ]\n",
        "\n",
        "all.equal(rownames(counts.mp.tf.disc), rownames(gene_annot_mp.tf_disc))  # Should return TRUE if aligned\n",
        "dim(counts.mp.tf.disc)\n",
        "dim(gene_annot_mp.tf_disc)\n",
        "View(gene_annot_mp.tf_disc)\n",
        "\n",
        "#In rtni analysis it needs \"SYMBOL\" in rowAnnotation\n",
        "colnames(gene_annot_mp.tf_disc)\n",
        "colnames(gene_annot_mp.tf_disc)[colnames(gene_annot_mp.tf_disc) == \"external_gene_name\"] <- \"SYMBOL\"\n",
        "colnames(gene_annot_mp.tf_disc)[colnames(gene_annot_mp.tf_disc) == \"ensembl_gene_id\"] <- \"ENSEMBL\"\n",
        "\n",
        "# Verify the change\n",
        "colnames(gene_annot_mp.tf_disc)\n",
        "\n",
        "\n",
        "# One final check:\n",
        "stopifnot(rownames(gene_annot_mp.tf_disc) == rownames(counts.mp.tf.disc), # features\n",
        "          rownames(sampleAnnotation.disc) == colnames(counts.mp.tf.disc)) # samples\n",
        "\n",
        "print(\"Step 5 completed: GeneAnnotation/RowAnnotation done for discovery set\")\n",
        "\n",
        "#---6.Create disc_mp.tf list/SummarizedExperiment-----------------------\n",
        "counts.mp.tf.disc <- as.matrix(counts.mp.tf.disc)\n",
        "View(counts.mp.tf.disc)\n",
        "str(counts.mp.tf.disc)\n",
        "\n",
        "disc_mp.tf <- list(\n",
        "  expData = counts.mp.tf.disc,                       # expData as an assay\n",
        "  rowAnnotation = gene_annot_mp.tf_disc,             # Gene annotations (row metadata)\n",
        "  colAnnotation = sampleAnnotation.disc           # Sample annotations (column metadata)\n",
        ")\n",
        "\n",
        "View(disc_mp.tf)\n",
        "###DEBUGGING ERRORS###\n",
        "#> rtni_disc_mp.tf.NUM <- tni.constructor(expData = disc_mp.tf.NUM$expData,\n",
        "#                                     +                              regulatoryElements = regulatoryElements,\n",
        "#                                     +                              rowAnnotation = disc_mp.tf.NUM$rowAnnotation,\n",
        "#                                     +                              colAnnotation = disc_mp.tf.NUM$colAnnotation)\n",
        "#-Preprocessing for input data...\n",
        "#Error: 'expData' should be a numeric matrix with genes on rows and\n",
        "#samples on cols!\n",
        "\n",
        "#Values are not numeric\n",
        "\n",
        "# Find elements that will turn into NA when coerced to numeric\n",
        "non_numeric_elements <- counts.mp.tf.disc[is.na(as.numeric(counts.mp.tf.disc))]\n",
        "\n",
        "# Display non-numeric elements to understand what they are\n",
        "print(non_numeric_elements)\n",
        "\n",
        "#---TO RETURN NUMERIC-BUT IT GIVES THE ERROR:\n",
        "#Warning message:\n",
        "#  In matrix(as.numeric(counts.mp.tf.disc), nrow = nrow(counts.mp.tf.disc),  :\n",
        "#             NAs introduced by coercion\n",
        "\n",
        "#SummarizedExperiment of rtni_disc_mp.tf.NUM_mp.tf_se\n",
        "#rtni_disc_mp.tf.NUM_mp.tf_se <- SummarizedExperiment(\n",
        "#  assays = counts.mp.tf.disc,\n",
        "#  rowData = gene_annot_mp.tf_disc,\n",
        "#  colData = sampleAnnotation.disc\n",
        "#)\n",
        "\n",
        "#View(rtni_disc_mp.tf.NUM_mp.tf_se)\n",
        "\n",
        "# Check dimensions and alignment\n",
        "#dim(counts.mp.tf.disc.NUM)\n",
        "\n",
        "#dim(disc_mp.tf.NUM_se$expData)  # Should match dimensions of counts.mp.tf.disc\n",
        "all.equal(colnames(disc_mp.tf.NUM$expData), rownames(sampleAnnotation.disc))  # Should return TRUE\n",
        "all.equal(disc_mp.tf.NUM$expData, counts.mp.tf.disc)\n",
        "all.equal(disc_mp.tf.NUM$colAnnotation, sampleAnnotation.disc)\n",
        "\n",
        "all.equal(disc_mp.tf$rowAnnotation,disc_mp.tf.NUM$rowAnnotation)\n",
        "#[1] TRUE\n",
        "all.equal(disc_mp.tf$colAnnotation,disc_mp.tf.NUM$colAnnotation)\n",
        "#[1] TRUE\n",
        "all.equal(disc_mp.tf$expData,disc_mp.tf.NUM$expData)\n",
        "#[1] \"Modes: character, numeric\"\n",
        "#[2] \"'is.NA' value mismatch: 14 in current 0 in target\"\n",
        "###DEBUGGING ERRORS###\n",
        "\n",
        "print(\"Step 6 completed: list/SummarizedExperiment file prepared for Discovery\")\n",
        "\n",
        "#---7.RegulatoryElements-----------------\n",
        "\n",
        "# Load TF annotation\n",
        "data(\"tfsData\")\n",
        "\n",
        "# Check TF annotation:\n",
        "# Intersect TFs from Lambert et al. (2018) with gene annotation\n",
        "# from the gene expression of 90 metabolic pathway cohort\n",
        "regulatoryElements <- intersect(tfsData$Lambert2018$SYMBOL, disc_mp.tf$rowAnnotation$SYMBOL)\n",
        "View(regulatoryElements)\n",
        "regulatoryElements <- sort(regulatoryElements)\n",
        "View(regulatoryElements)\n",
        "\n",
        "print(\"Step 7 completed: RegulatoryElements Defined\")\n",
        "\n",
        "#---8.Run the TNI constructor with the extracted matrix for disc_mp.tf.NUM--------------\n",
        "#This dataset consists of a list with 3 objects:\n",
        "##a named gene expression matrix (tniData$expData),\n",
        "##a data frame with gene annotations (tniData$rowAnnotation),\n",
        "##and a data frame with sample annotations (tniData$colAnnotation).\n",
        "##alternatively, 'expData' can be a 'SummarizedExperiment' object\n",
        "rtni_disc_mp.tf <- tni.constructor(expData = disc_mp.tf$expData,\n",
        "                             regulatoryElements = regulatoryElements,\n",
        "                             rowAnnotation = disc_mp.tf$rowAnnotation,\n",
        "                             colAnnotation = disc_mp.tf$colAnnotation)\n",
        "\n",
        "save(rtni_disc_mp.tf, file = file.path(\"~/NCA.ER/NCA.MP.TF.DISC.NCA/\", \"rtni_mp.tf.disc.RData\"))\n",
        "\n",
        "###snow clustering returns zero results\n",
        "### Compute the reference regulatory network by permutation and bootstrap analyses.\n",
        "### Please set 'spec' according to your available hardware\n",
        "#options(cluster=snow::makeCluster(spec=4, \"SOCK\"))  #???\n",
        "\n",
        "#1.with pValuCutoff =1e-7 it gives 0 values, even in the summary\n",
        "#rtni_disc_mp.tf.NUM <- tni.permutation(rtni_disc_mp.tf.NUM, pValueCutoff = 1e-7) #pValueCutoff=1e-7 ?? zero_values\n",
        "\n",
        "#2.with pValuCutoff= 1e-7 and npermutation = 1000 it gives 0 values, even in the sumary\n",
        "#rtni_disc_mp.tf.NUM <- tni.permutation(rtni_disc_mp.tf.NUM,nPermutations = 1000, pValueCutoff = 1e-7) #pValueCutoff=1e-7 ?? zero_values\n",
        "\n",
        "#3.  nPermutations >= 1000 with running snow goes the same zero values\n",
        "#4.  nPermutations >= 1000 without snow package, and pValueCutoff works!\n",
        "rtni_disc_mp.tf <- tni.permutation(rtni_disc_mp.tf, nPermutations = 1000)\n",
        "\n",
        "#Unstable interactions are subsequently removed by bootstrap analysis,\n",
        "##creates a consensus bootstrap network, referred here as refnet (reference network).\n",
        "rtni_disc_mp.tf <- tni.bootstrap(rtni_disc_mp.tf)\n",
        "\n",
        "#stopCluster(getOption(\"cluster\"))  # avoid leak memory\n",
        "\n",
        "# Compute the DPI-filtered regulatory network\n",
        "rtni_disc_mp.tf <- tni.dpi.filter(rtni_disc_mp.tf, eps = NA)\n",
        "tni.regulon.summary(rtni_disc_mp.tf)\n",
        "\n",
        "#detailed information about a specific regulon\n",
        "tni.regulon.summary(rtni_disc_mp.tf, regulatoryElements = \"SATB1\")\n",
        "\n",
        "# Save the TNI object for subsequent analyses\n",
        "save(rtni_disc_mp.tf, file = file.path(\"~/NCA.ER/NCA.MP.TF.DISC.NCA/\", \"rtni_disc_mp.tf.RData\"))\n",
        "\n",
        "print(\"Step 8 completed: Loaded rtni data and saved.\")\n",
        "\n",
        "regulons <- tni.get(rtni_disc_mp.tf, what = \"regulons.and.mode\", idkey = \"SYMBOL\")\n",
        "View(regulons)\n",
        "head(regulons)\n",
        "\n",
        "write.table(regulons$BCL11A,file = file.path(\"~/NCA.ER/NCA.MP.TF.DISC.NCA/Regulons/\",\"BCL11A.Regulon.txt\"),sep = \"\\t\")\n",
        "# Load necessary library to get all the regulons\n",
        "library(dplyr)\n",
        "\n",
        "# Assuming `regulons` is a list where each element is named by the regulon (e.g., `regulons$DNMT1`)\n",
        "for (regulon_name in names(regulons)) {\n",
        "  regulon_data <- regulons[[regulon_name]]  # Access each regulon as a vector\n",
        "\n",
        "  # Check if the regulon has more than one gene\n",
        "  if (length(regulon_data) > 1) {\n",
        "    # Prepare a data frame for output, including the gene names and values\n",
        "    output_df <- data.frame(Gene = names(regulon_data), Value = regulon_data)\n",
        "\n",
        "    # Prepare the output file path\n",
        "    output_file <- paste0(\"~/NCA.ER/NCA.MP.TF.DISC.NCA/Regulons/\", regulon_name, \".DISC.txt\")\n",
        "\n",
        "    # Write the regulon data to a file\n",
        "    write.table(output_df, file = output_file, sep = \"\\t\", quote = FALSE, row.names = FALSE, col.names = FALSE)\n",
        "  }\n",
        "}\n",
        "\n",
        "    ###To extract regulons all in one txt.###\n",
        "# Find the maximum number of genes across all regulons\n",
        "max_genes <- max(sapply(regulon.NA, function(x) if (is.null(x)) 0 else length(x)))\n",
        "\n",
        "# Create a list of data frames with equal row lengths\n",
        "regulon_list <- lapply(names(regulon.NA), function(regulon_name) {\n",
        "  regulon_data <- regulon.NA[[regulon_name]]\n",
        "\n",
        "  # Handle empty or NULL regulons\n",
        "  if (is.null(regulon_data) || length(regulon_data) == 0) {\n",
        "    df <- data.frame(\n",
        "      Gene = rep(NA, max_genes),\n",
        "      Value = rep(NA, max_genes),\n",
        "      stringsAsFactors = FALSE\n",
        "    )\n",
        "  } else if (is.vector(regulon_data)) {\n",
        "    # Ensure matching lengths for Gene and Value\n",
        "    genes <- names(regulon_data)\n",
        "    values <- regulon_data\n",
        "    if (length(genes) == 0) genes <- rep(NA, length(values))\n",
        "    df <- data.frame(\n",
        "      Gene = genes,\n",
        "      Value = values,\n",
        "      stringsAsFactors = FALSE\n",
        "    )\n",
        "  } else if (is.matrix(regulon_data) || is.data.frame(regulon_data)) {\n",
        "    df <- data.frame(\n",
        "      Gene = rownames(regulon_data),\n",
        "      Value = regulon_data[, 1], # Assuming values are in the first column\n",
        "      stringsAsFactors = FALSE\n",
        "    )\n",
        "  } else {\n",
        "    df <- data.frame(\n",
        "      Gene = rep(NA, max_genes),\n",
        "      Value = rep(NA, max_genes),\n",
        "      stringsAsFactors = FALSE\n",
        "    )\n",
        "  }\n",
        "\n",
        "  # Extend to max_genes rows if needed\n",
        "  if (nrow(df) < max_genes) {\n",
        "    df <- rbind(df, data.frame(\n",
        "      Gene = rep(NA, max_genes - nrow(df)),\n",
        "      Value = rep(NA, max_genes - nrow(df))\n",
        "    ))\n",
        "  }\n",
        "\n",
        "  # Rename columns with regulon names\n",
        "  colnames(df) <- c(paste0(regulon_name, \"_Gene\"), paste0(regulon_name, \"_Value\"))\n",
        "  return(df)\n",
        "})\n",
        "\n",
        "# Combine all into one data frame\n",
        "regulon_df <- do.call(cbind, regulon_list)\n",
        "\n",
        "# Write to file\n",
        "write.table(regulon_df, file = file.path(\"~/NCA.ER/NCA.MP.TF.Finalized/NCA.Disc.MP.tf/\", \"Disc.regulon.NA.txt\"), sep = \"\\t\", row.names = FALSE, quote = FALSE)\n",
        "\n",
        "print(\"Step 9 completed: Regulons Extracted\")\n",
        "#---9. Regulon activity profiles----------\n",
        "\n",
        "library(Fletcher2013b)\n",
        "library(pheatmap)\n",
        "library(grid)\n",
        "library(gridExtra)\n",
        "\n",
        "# Load 'rtni1st' data object, which includes regulons and expression profiles\n",
        "#here I will use: rtni_disc_mp.tf.NUM\n",
        "\n",
        "# A list of transcription factors of interest ( #9 )\n",
        "# Compute regulon activity for individual samples\n",
        "rtni_disc_mp.tf <- tni.gsea2(rtni_disc_mp.tf, regulatoryElements = rtni_disc_mp.tf@regulatoryElements)\n",
        "metabric_regact_disc <- tni.get(rtni_disc_mp.tf, what = \"regulonActivity\")\n",
        "View(metabric_regact_disc)\n",
        "\n",
        "# Get sample attributes from the 'rtni_disc_mp.tf.NUM' dataset\n",
        "metabric_annot_disc <- tni.get(rtni_disc_mp.tf, \"colAnnotation\")\n",
        "\n",
        "# Get ER+/- and PAM50 attributes for pheatmap\n",
        "attribs_disc <- c(\"ER+\",\"ER-\")\n",
        "metabric_annot_disc <- metabric_annot_disc[,attribs_disc]\n",
        "\n",
        "save(rtni_disc_mp.tf, file = file.path(\"~/NCA.ER/NCA.MP.TF.DISC.NCA/\", \"rtni_disc_mp.tf.gsea.RData\"))\n",
        "\n",
        "# Step 1: Identify samples with ER+ = 1\n",
        "ER_positive_samples <- rownames(metabric_annot_disc)[metabric_annot_disc$`ER+` == 1]\n",
        "ER_negative_samples <- rownames(metabric_annot_disc)[metabric_annot_disc$`ER+` == 0]\n",
        "\n",
        "# Step 2: Order columns in metabric_regact_disc with ER+ samples on the left\n",
        "ordered_sample_names <- c(ER_positive_samples, ER_negative_samples)\n",
        "metabric_regact_disc_ordered <- metabric_regact_disc$differential[ordered_sample_names,]\n",
        "\n",
        "# Define custom colors for each category in annotation_col\n",
        "ER_annotation <- metabric_annot_disc[,c(\"ER+\",\"ER-\")]\n",
        "ER_annotation_colors <- list(\n",
        "  \"ER+\" = c(\"0\" = \"lightgrey\", \"1\" = \"blue\"),\n",
        "  \"ER-\" = c(\"0\" = \"lightgrey\", \"1\" = \"red\")\n",
        "  )\n",
        "\n",
        "pdf(\"~/NCA.ER/NCA.MP.TF.DISC.NCA/Disc.NCA.MP.TF.Heatmap.pdf\", width = 10, height = 10)\n",
        "\n",
        "# Plot regulon activity profiles\n",
        "disc.heatmap <- pheatmap(t(metabric_regact_disc_ordered),\n",
        "                         main=\"Discovery Set (n=988 samples)\",\n",
        "                         annotation_col = ER_annotation,\n",
        "                         show_colnames = FALSE, annotation_legend = FALSE,\n",
        "                         clustering_method = \"ward.D2\", fontsize_row = 3,\n",
        "                         clustering_distance_rows = \"correlation\",\n",
        "                         cluster_cols = FALSE,\n",
        "                         legend = TRUE,\n",
        "                         annotation_colors = ER_annotation_colors,\n",
        "                         fontsize_col = 3, fontsize = 6,\n",
        "                         border_color = NA,\n",
        ")\n",
        "\n",
        "grid.text(\"Regulons\", x= 0.97 , y=0.3, rot=270)\n",
        "\n",
        "dev.off()\n",
        "\n",
        "disc.session.info <- sessionInfo()\n",
        "\n",
        "#---10.Transcriptional Network Analysis (TNA)-----------\n",
        "Metabric_Manual_disc_tna <- read.delim(\"~/NCA/data/METABRIC Clinical.txt\", row.names = 1)\n",
        "View(Metabric_Manual_disc_tna)\n",
        "\n",
        "Metabric_Manual_disc_tna <- Metabric_Manual_disc_tna[order(rownames(Metabric_Manual_disc_tna)),]\n",
        "View(Metabric_Manual_disc_tna)\n",
        "\n",
        "# Replace dots with underscores in column names if needed\n",
        "rownames(Metabric_Manual_disc_tna) <- gsub(\"\\\\-\", \"_\", rownames(Metabric_Manual_disc_tna))\n",
        "View(Metabric_Manual_disc_tna)\n",
        "\n",
        "disc_sample_mp.tf_ids_tna <- colnames(disc.mp.tf)\n",
        "View(disc_sample_mp.tf_ids_tna)\n",
        "length(disc_sample_mp.tf_ids_tna)\n",
        "\n",
        "sort(disc_sample_mp.tf_ids_tna)\n",
        "View(disc_sample_mp.tf_ids_tna)\n",
        "length(disc_sample_mp.tf_ids_tna)\n",
        "\n",
        "all.equal(disc_sample_mp.tf_ids_tna,colnames(disc.mp.tf))\n",
        "\n",
        "# Find the common sample IDs between the two datasets\n",
        "common_samples_disc_tna <- intersect(rownames(Metabric_Manual_disc_tna), disc_sample_mp.tf_ids_tna)\n",
        "common_samples_disc_tna <- sort(common_samples_disc_tna)\n",
        "View(common_samples_disc_tna)\n",
        "length(common_samples_disc_tna)\n",
        "\n",
        "sampleAnnotation.disc_tna <- Metabric_Manual_disc_tna[common_samples_disc_tna, ]\n",
        "dim(sampleAnnotation.disc_tna)\n",
        "\n",
        "# Verify that they are aligned\n",
        "all.equal(colnames(counts.mp.tf.disc), rownames(sampleAnnotation.disc_tna))  # Should return TRUE\n",
        "\n",
        "# If the first is not TRUE, you can match up the samples/columns in\n",
        "# counts with the samples/rows in sampleAnnotation.disc like this (which is fine\n",
        "# to run even if the first was TRUE):\n",
        "\n",
        "#tempindex <- match(colnames(counts.mp.tf.disc), rownames(sample_metadata))\n",
        "#sampleAnnotation.disc <- sample_metadata[tempindex, ]\n",
        "\n",
        "#Check again\n",
        "\n",
        "all.equal(colnames(counts.mp.tf.disc), rownames(sampleAnnotation.disc_tna))\n",
        "all(disc_mp.tf$colAnnotation$IDs == rownames(sampleAnnotation.disc_tna))\n",
        "\n",
        "# Check that row names of sample_annotations match column names of expression_data\n",
        "if (!all(rownames(sampleAnnotation.disc_tna) == colnames(disc_mp.tf$expData))) {\n",
        "  stop(\"Mismatch between sample annotation rownames and expression data colnames!\")\n",
        "}\n",
        "\n",
        "# Load required package\n",
        "library(limma)\n",
        "\n",
        "# Create the design matrix for the linear model\n",
        "# Assuming your label column is named \"ER_status\" with values \"ERpos\" and \"ERneg\"\n",
        "design <- model.matrix(~ 0 + factor(sampleAnnotation.disc_tna$ER.Status))\n",
        "colnames(design) <- levels(factor(sampleAnnotation.disc_tna$ER.Status))\n",
        "rownames(design) <- rownames(sampleAnnotation.disc_tna)\n",
        "View(design)\n",
        "\n",
        "all.equal(as.vector(design[,\"Positive\"]),as.vector(sampleAnnotation.disc$`ER+`))\n",
        "all(design[,\"Positive\"] == sampleAnnotation.disc$`ER+`)\n",
        "all(design[,\"Negative\"] == sampleAnnotation.disc$`ER-`)\n",
        "\n",
        "# Fit the linear model using limma\n",
        "fit <- lmFit(counts.mp.tf.disc, design)\n",
        "\n",
        "#---10.1_Pos Vs Neg--------------------------\n",
        "# Create contrast matrix to compare ERpos vs ERneg\n",
        "contrast_matrix <- makeContrasts(Positive_vs_Negative = Positive - Negative, levels = design)\n",
        "\n",
        "# Apply the contrast matrix\n",
        "fit2 <- contrasts.fit(fit, contrast_matrix)\n",
        "\n",
        "# Empirical Bayes adjustment\n",
        "fit2 <- eBayes(fit2)\n",
        "\n",
        "\n",
        "# Extract results (log2 fold changes, p-values, etc.)\n",
        "phenotype <- topTable(fit2, coef = \"Positive_vs_Negative\", adjust.method = \"BH\", number = Inf)\n",
        "write.table(phenotype,file = file.path(\"~/NCA.ER/NCA.MP.TF.DISC.NCA/\",\"DEG.Disc.Pos.Vs.Neg.txt\"),sep = \"\\t\")\n",
        "# Save results to a file if needed\n",
        "write.table(phenotype, file = \"differential_expression_results.txt\", sep = \"\\t\", quote = FALSE, row.names = TRUE)\n",
        "\n",
        "# Output the top results for inspection\n",
        "View(phenotype)\n",
        "dim(phenotype)\n",
        "\n",
        "# Order phenotype alphabetically by row names\n",
        "phenotype <- phenotype[order(rownames(phenotype)), ]\n",
        "\n",
        "all(rownames(phenotype) == rownames(counts.mp.tf.disc))\n",
        "all(rownames(phenotype) == gene_annot_mp.tf_disc$SYMBOL)\n",
        "\n",
        "# Extract 'logFC' as a named numeric vector\n",
        "logFC_disc_mp.tf <- setNames(phenotype$logFC, rownames(phenotype))\n",
        "View(logFC_disc_mp.tf)\n",
        "\n",
        "all.equal(names(logFC_disc_mp.tf),rownames(phenotype))\n",
        "all.equal(as.vector(logFC_disc_mp.tf),phenotype$logFC)\n",
        "\n",
        "# Filter genes with significant adjusted p-values and a logFC threshold\n",
        "hits <- subset(phenotype, adj.P.Val < 0.05 & abs(logFC) > 1)\n",
        "\n",
        "# View the differentially expressed genes\n",
        "View(hits)\n",
        "dim(hits)\n",
        "\n",
        "all(rownames(phenotype) == gene_annot_mp.tf_disc$SYMBOL)\n",
        "\n",
        "#Error: NOTE: all names in 'phenotype' should be available in col1 of 'phenoIDs'!\n",
        "library(dplyr)\n",
        "gene_annot_mp.tf_disc_tna <- gene_annot_mp.tf_disc %>%\n",
        "  select(SYMBOL, everything())\n",
        "\n",
        "View(gene_annot_mp.tf_disc_tna)\n",
        "all(gene_annot_mp.tf_disc$SYMBOL == gene_annot_mp.tf_disc_tna$SYMBOL)\n",
        "all(gene_annot_mp.tf_disc$ENSEMBL == gene_annot_mp.tf_disc_tna$ENSEMBL)\n",
        "\n",
        "tna.disc_mp.tf <- list(\n",
        "  phenotype = logFC_disc_mp.tf,\n",
        "  phenoID = gene_annot_mp.tf_disc_tna,\n",
        "  hits = rownames(hits)\n",
        ")\n",
        "\n",
        "View(tna.disc_mp.tf)\n",
        "# Input 1: 'object', a TNI object with regulons\n",
        "# Input 2: 'phenotype', a named numeric vector, usually log2 differential expression levels\n",
        "# Input 3: 'hits', a character vector, usually a set of differentially expressed genes\n",
        "# Input 4: 'phenoIDs', an optional data frame with gene anottation mapped to the phenotype\n",
        "\n",
        "#set.seed(123)  Since I didnt run everything from scratch for this part I used set.seed() in saved Rproject\n",
        "\n",
        "#CHECK \"rtnaData\"\n",
        "rtna_disc.mp.tf <- tni2tna.preprocess(object = rtni_disc_mp.tf,\n",
        "                           phenotype = tna.disc_mp.tf$phenotype,\n",
        "                           hits = tna.disc_mp.tf$hits,\n",
        "                           phenoIDs = tna.disc_mp.tf$phenoID)\n",
        "# Run the MRA method\n",
        "rtna_disc.mp.tf <- tna.mra(rtna_disc.mp.tf)\n",
        "\n",
        "# Get MRA results;\n",
        "#..setting 'ntop = -1' will return all results, regardless of a threshold\n",
        "mra_disc.mp.tf <- tna.get(rtna_disc.mp.tf, what=\"mra\", ntop = -1)\n",
        "View(mra_disc.mp.tf)\n",
        "\n",
        "# Run the GSEA method\n",
        "# Please set nPermutations >= 1000\n",
        "rtna_disc.mp.tf <- tna.gsea1(rtna_disc.mp.tf, nPermutations=1000)\n",
        "\n",
        "# Get GSEA results\n",
        "gsea1_disc.mp.tf <- tna.get(rtna_disc.mp.tf, what=\"gsea1\", ntop = -1)\n",
        "head(gsea1_disc.mp.tf)\n",
        "View(gsea1_disc.mp.tf)\n",
        "\n",
        "# Filter for significant TFs\n",
        "gsea1_disc.mp.tf.sig <- gsea1_disc.mp.tf[gsea1_disc.mp.tf$Adjusted.Pvalue <= 0.05, ]\n",
        "#View(gsea1_disc.mp.tf.sig)\n",
        "\n",
        "# Specify TFs to include in the plot\n",
        "sig_tfs_disc_mp.tf <- rownames(gsea1_disc.mp.tf.sig)\n",
        "\n",
        "# Plot GSEA results for significant TFs\n",
        "tna.plot.gsea1(\n",
        "  rtna_disc.mp.tf,\n",
        "  labPheno = \"abs(log2 fold changes)\",\n",
        "  tfs = sig_tfs_disc_mp.tf, # Include only significant TFs\n",
        "  file = \"gsea1_plot_disc_mp.tf_sig.pdf\",\n",
        "  filepath = \"~/NCA.ER/NCA.MP.TF.DISC.NCA/\"\n",
        ")\n",
        "\n",
        "# Plot GSEA results\n",
        "tna.plot.gsea1(rtna_disc.mp.tf,\n",
        "               labPheno=\"abs(log2 fold changes)\",\n",
        "               ntop = 5,\n",
        "               file = \"gsea1_plot_top5\",\n",
        "               ylimPanels = c(0.0,3,0.0,0.0),\n",
        "               heightPanels = c(1,1,3))\n",
        "\n",
        "\n",
        "# Run the GSEA-2T method\n",
        "# Please set nPermutations >= 1000\n",
        "rtna_disc.mp.tf <- tna.gsea2(rtna_disc.mp.tf, nPermutations = 1000)\n",
        "# Get GSEA-2T results\n",
        "gsea2_disc.mp.tf <- tna.get(rtna_disc.mp.tf, what = \"gsea2\", ntop = -1)\n",
        "head(gsea2_disc.mp.tf$differential)\n",
        "\n",
        "# Plot GSEA-2T results\n",
        "tna.plot.gsea2(rtna_disc.mp.tf, labPheno=\"log2 fold changes\", tfs=\"SATB1\", file = \"Plot_disc_mp.tf\")\n",
        "#---10.2---Neg Vs Pos--------------------\n",
        "# Create contrast matrix to compare ERpos vs ERneg\n",
        "contrast_matrix_neg.Vs.Pos <- makeContrasts(Negative_vs_Positive = Negative - Positive, levels = design)\n",
        "\n",
        "# Apply the contrast matrix\n",
        "fit2_neg.Vs.Pos <- contrasts.fit(fit, contrast_matrix_neg.Vs.Pos)\n",
        "View(fit2_neg.Vs.Pos)\n",
        "\n",
        "# Empirical Bayes adjustment\n",
        "fit2_neg.Vs.Pos <- eBayes(fit2_neg.Vs.Pos)\n",
        "\n",
        "\n",
        "# Extract results (log2 fold changes, p-values, etc.)\n",
        "phenotype_neg.Vs.Pos <- topTable(fit2_neg.Vs.Pos, coef = \"Negative_vs_Positive\", adjust.method = \"BH\", number = Inf)\n",
        "\n",
        "# Save results to a file if needed\n",
        "write.table(phenotype_neg.Vs.Pos, file = \"differential_expression_results_neg.Vs.Pos.txt\", sep = \"\\t\", quote = FALSE, row.names = TRUE)\n",
        "\n",
        "# Output the top results for inspection\n",
        "View(phenotype_neg.Vs.Pos)\n",
        "dim(phenotype_neg.Vs.Pos)\n",
        "View(phenotype)\n",
        "\n",
        "# Order phenotype alphabetically by row names\n",
        "phenotype_neg.Vs.Pos <- phenotype_neg.Vs.Pos[order(rownames(phenotype_neg.Vs.Pos)), ]\n",
        "\n",
        "all(rownames(phenotype_neg.Vs.Pos) == rownames(counts.mp.tf.disc))\n",
        "all(rownames(phenotype_neg.Vs.Pos) == gene_annot_mp.tf_disc$SYMBOL)\n",
        "\n",
        "write.table(phenotype_neg.Vs.Pos,file = file.path(\"~/NCA.ER/NCA.MP.TF.DISC.NCA/\",\"DEG.Disc.Neg.Vs.Pos.txt\"),sep = \"\\t\")\n",
        "\n",
        "# Extract 'logFC' as a named numeric vector\n",
        "logFC_disc_mp.tf_neg.Vs.Pos <- setNames(phenotype_neg.Vs.Pos$logFC, rownames(phenotype_neg.Vs.Pos))\n",
        "View(logFC_disc_mp.tf_neg.Vs.Pos)\n",
        "\n",
        "all.equal(names(logFC_disc_mp.tf_neg.Vs.Pos),rownames(phenotype_neg.Vs.Pos))\n",
        "all.equal(as.vector(logFC_disc_mp.tf_neg.Vs.Pos),phenotype_neg.Vs.Pos$logFC)\n",
        "\n",
        "# Filter genes with significant adjusted p-values and a logFC threshold\n",
        "hits_neg.Vs.Pos <- subset(phenotype_neg.Vs.Pos, adj.P.Val < 0.05 & abs(logFC_disc_mp.tf_neg.Vs.Pos) > 1)\n",
        "\n",
        "# View the differentially expressed genes\n",
        "View(hits_neg.Vs.Pos)\n",
        "dim(hits_neg.Vs.Pos)\n",
        "\n",
        "all(rownames(phenotype_neg.Vs.Pos) == gene_annot_mp.tf_disc_tna$SYMBOL)\n",
        "\n",
        "#Error: NOTE: all names in 'phenotype' should be available in col1 of 'phenoIDs'!\n",
        "library(dplyr)\n",
        "\n",
        "all(gene_annot_mp.tf_disc_tna$SYMBOL == rownames(phenotype_neg.Vs.Pos))\n",
        "all(rownames(phenotype) == rownames(phenotype_neg.Vs.Pos))\n",
        "\n",
        "#phenoID remained the same\n",
        "\n",
        "tna.disc_mp.tf_neg.Vs.Pos <- list(\n",
        "  phenotype = logFC_disc_mp.tf_neg.Vs.Pos,\n",
        "  phenoID = gene_annot_mp.tf_disc_tna,\n",
        "  hits = rownames(hits_neg.Vs.Pos)\n",
        ")\n",
        "\n",
        "View(tna.disc_mp.tf_neg.Vs.Pos)\n",
        "# Input 1: 'object', a TNI object with regulons\n",
        "# Input 2: 'phenotype', a named numeric vector, usually log2 differential expression levels\n",
        "# Input 3: 'hits', a character vector, usually a set of differentially expressed genes\n",
        "# Input 4: 'phenoIDs', an optional data frame with gene anottation mapped to the phenotype\n",
        "\n",
        "#set.seed(123)  Since I didnt run everything from scratch for this part I used set.seed() in saved Rproject\n",
        "\n",
        "#CHECK \"rtnaData\"\n",
        "rtna_disc.mp.tf_neg.Vs.Pos <- tni2tna.preprocess(object = rtni_disc_mp.tf,\n",
        "                                      phenotype = tna.disc_mp.tf_neg.Vs.Pos$phenotype,\n",
        "                                      hits = tna.disc_mp.tf_neg.Vs.Pos$hits,\n",
        "                                      phenoIDs = tna.disc_mp.tf_neg.Vs.Pos$phenoID)\n",
        "# Run the MRA method\n",
        "rtna_disc.mp.tf_neg.Vs.Pos <- tna.mra(rtna_disc.mp.tf_neg.Vs.Pos)\n",
        "\n",
        "# Get MRA results;\n",
        "#..setting 'ntop = -1' will return all results, regardless of a threshold\n",
        "mra_disc.mp.tf_neg.Vs.Pos <- tna.get(rtna_disc.mp.tf_neg.Vs.Pos, what=\"mra\", ntop = -1)\n",
        "View(mra_disc.mp.tf_neg.Vs.Pos)\n",
        "\n",
        "all(mra_disc.mp.tf == mra_disc.mp.tf_neg.Vs.Pos)\n",
        "\n",
        "# Run the GSEA method\n",
        "# Please set nPermutations >= 1000\n",
        "rtna_disc.mp.tf_neg.Vs.Pos <- tna.gsea1(rtna_disc.mp.tf_neg.Vs.Pos, nPermutations=1000)\n",
        "\n",
        "# Get GSEA results\n",
        "gsea1_disc.mp.tf_neg.Vs.Pos <- tna.get(rtna_disc.mp.tf_neg.Vs.Pos, what=\"gsea1\", ntop = -1)\n",
        "head(gsea1_disc.mp.tf_neg.Vs.Pos)\n",
        "View(gsea1_disc.mp.tf_neg.Vs.Pos)\n",
        "\n",
        "# Filter for significant TFs\n",
        "gsea1_disc.mp.tf.sig_neg.Vs.Pos <- gsea1_disc.mp.tf_neg.Vs.Pos[gsea1_disc.mp.tf_neg.Vs.Pos$Adjusted.Pvalue <= 0.05, ]\n",
        "View(gsea1_disc.mp.tf.sig_neg.Vs.Pos)\n",
        "\n",
        "all(gsea1_disc.mp.tf.sig_neg.Vs.Pos == gsea1_disc.mp.tf.sig)\n",
        "#Error in Ops.data.frame(gsea1_disc.mp.tf.sig_neg.Vs.Pos, gsea1_disc.mp.tf.sig) :\n",
        "#‘==’ only defined for equally-sized data frames\n",
        "\n",
        "# Specify TFs to include in the plot\n",
        "sig_tfs_disc_mp.tf_neg.Vs.Pos <- rownames(gsea1_disc.mp.tf.sig_neg.Vs.Pos)\n",
        "\n",
        "# Plot GSEA results for significant TFs\n",
        "tna.plot.gsea1(\n",
        "  rtna_disc.mp.tf_neg.Vs.Pos,\n",
        "  labPheno = \"abs(log2 fold changes)\",\n",
        "  tfs = sig_tfs_disc_mp.tf_neg.Vs.Pos, # Include only significant TFs\n",
        "  file = \"gsea1_plot_disc_mp.tf_sig_neg.Vs.Pos.pdf\",\n",
        "  filepath = \"~/NCA.ER/NCA.MP.TF.DISC.NCA/\"\n",
        ")\n",
        "\n",
        "# Plot GSEA results\n",
        "tna.plot.gsea1(rtna_disc.mp.tf_neg.Vs.Pos,\n",
        "               labPheno=\"abs(log2 fold changes)\",\n",
        "               ntop = 5,\n",
        "               file = \"gsea1_plot_neg.Vs.Pos_top5\",\n",
        "               ylimPanels = c(0.0,3,0.0,0.0),\n",
        "               heightPanels = c(1,1,3))\n",
        "\n",
        "\n",
        "# Run the GSEA-2T method\n",
        "# Please set nPermutations >= 1000\n",
        "rtna_disc.mp.tf_neg.Vs.Pos <- tna.gsea2(rtna_disc.mp.tf_neg.Vs.Pos, nPermutations = 1000)\n",
        "# Get GSEA-2T results\n",
        "gsea2_disc.mp.tf_neg.Vs.Pos <- tna.get(rtna_disc.mp.tf_neg.Vs.Pos, what = \"gsea2\", ntop = -1)\n",
        "head(gsea2_disc.mp.tf_neg.Vs.Pos$differential)\n",
        "\n",
        "# Plot GSEA-2T results\n",
        "tna.plot.gsea2(rtna_disc.mp.tf_neg.Vs.Pos, labPheno=\"log2 fold changes\", tfs=\"GATA3\", file = \"Plot_disc_mp_neg.Vs.Pos.tf\")\n",
        "\n",
        "#---11.Extract Target Genes of Each Regulon-------------------------\n",
        "\n",
        "writeLines(capture.output(sessionInfo()), \"sessionInfo_disc_mp.tf.txt\")"
      ],
      "metadata": {
        "id": "u7A38QY64j-2"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "name": "Welcome To Colab",
      "toc_visible": true,
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}