{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPEgaihnS7RV16gFLbzBB6M",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/kosarHNajafi/R.Codes/blob/main/RTN_NCA_Finalized.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        ">[1.Curating data set](#scrollTo=Oyrqhny5FFXs)\n",
        "\n",
        ">[2.Which Data to Use regarding Discovery Data set?](#scrollTo=otBerHwcSBjw)\n",
        "\n",
        ">[3.Sample Annotation(Column Annotation)](#scrollTo=T4C7sq_9cKyQ)\n",
        "\n",
        ">[4.Gene Annotation](#scrollTo=aiQreEAVnV4B)\n",
        "\n",
        ">[5.Build First rtni-data](#scrollTo=_HmzcQkbeXY3)\n",
        "\n",
        ">[6.Building Regulons](#scrollTo=xDCQnICdrkle)\n",
        "\n"
      ],
      "metadata": {
        "colab_type": "toc",
        "id": "bhnOihIlK8m3"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "#1.Curating data set\n",
        "\n",
        "\n",
        "> #Discovery Data Set(First Code cell)\n",
        "\n",
        "There are some  non_numeric_elements in Discovery data set;such as NA and \"5.14893783197 5.\", \"11.3176307652 11\" \"5.455753203 5.36\".\n",
        "\n",
        "before working with all 19220 gene expressions in discovery data set(which one of them is \"NA\" thus deleted and remained 19219 genes in the row; non_numeric values should be handled.\n",
        "\n",
        "\n",
        "> #Validation(Second Code Cell)\n",
        "\n",
        "In the validation data set there is no such problem"
      ],
      "metadata": {
        "id": "Oyrqhny5FFXs"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "F-PxXB4NE-bH"
      },
      "outputs": [],
      "source": [
        "#R File from '~/NCA.ER/NCA.METABRIC/METABRIC.NCA.MP.TF.EPSNA.R'\n",
        "#---Load libraries-----------------\n",
        "library(biomaRt)\n",
        "library(RTN)\n",
        "#---1.Load Gene Expression Data of 90 metabolic pathways and TFs = a mtrix for counts/assays---------------------------------\n",
        "#load txt. Format for available genes in the discovery dataset\n",
        "Disc.ALL <- read.delim(\"~/NCA.ER/data/Disc.Raw.Genes.all.txt\",header = TRUE, as.is = TRUE)\n",
        "\n",
        "sum(is.na(Disc.ALL[, 1]))  # Count the number of NA values\n",
        "#[1] 1\n",
        "\n",
        "which(is.na(Disc.ALL[, 1]))  # Identify their positions\n",
        "#[1] 11028 in the main file not sorted, only metabolic clusters row was removed.\n",
        "# [1] 10390 in the sorted data based on gene names without 1 and 0 metabolic clusters.\n",
        "\n",
        "Disc.ALL <- Disc.ALL[-11028,]  # Remove the row with 'NA' as gene name\n",
        "rownames(Disc.ALL) <- Disc.ALL[,1]\n",
        "Disc.ALL <- Disc.ALL[,-1]\n",
        "dim(Disc.ALL)\n",
        "#[1] 19219  993\n",
        "\n",
        "str(Disc.ALL)\n",
        "#'data.frame':\t19219 obs. of  993 variables:\n",
        "#$ MB.0005: chr  \"6.24393549\" \"5.228923847\" \"6.781157168\" \"5.583909882\" ...\n",
        "#$ MB.0006: num  6.22 5.91 6.75 5.36 6.08 ...\n",
        "#$ MB.0008: num  6.06 5.3 7.66 5.3 5.69 ...\n",
        "#$ MB.0014: num  6.16 5.4 6.68 5.85 5.65 ...\n",
        "\n",
        "View(Disc.ALL)\n",
        "\n",
        "colnames(Disc.ALL) <- gsub(\"\\\\.\", \"_\", colnames(Disc.ALL))\n",
        "\n",
        "#Sort data by columnnames(Sample IDs) and rownames(Gene IDs)\n",
        "Disc.ALL <- Disc.ALL[order(rownames(Disc.ALL)),order(colnames(Disc.ALL))]\n",
        "View(Disc.ALL)\n",
        "head(Disc.ALL)\n",
        "str(Disc.ALL)\n",
        "#'data.frame':\t19219 obs. of  993 variables:\n",
        "#$ MB_0005: chr  \"5.059348923\" \"9.924912584\" \"5.421198528\" \"6.24393549\" ...\n",
        "#$ MB_0006: num  5.37 11.31 5.14 6.22 5.91 ...\n",
        "#$ MB_0008: num  5.28 9.83 5.56 6.06 5.3 ...\n",
        "#$ MB_0014: num  5.53 10.68 5.36 6.16 5.4 ...\n",
        "#$ MB_0020: num  5.28 9.31 5.69 6.16 5.53 ...\n",
        "\n",
        "#check for NAs\n",
        "sum(is.na(Disc.ALL))\n",
        "#[1] 10\n",
        "\n",
        "which(is.na((rownames(Disc.ALL))))\n",
        "#integer(0)\n",
        "\n",
        "#check for non-numeric elements\n",
        "counts.Disc.all <- as.matrix(Disc.ALL)\n",
        "\n",
        "# Find elements that will turn into NA when coerced to numeric\n",
        "non_numeric_elements1 <- counts.Disc.all[is.na(as.numeric(counts.Disc.all))]\n",
        "#Warning message:\n",
        "#NAs introduced by coercion\n",
        "\n",
        "length(non_numeric_elements1)\n",
        "#[1] 204\n",
        "\n",
        "write.table(non_numeric_elements1,\"~/NCA.ER/NCA.Disc/non_numeric_all.txt\",sep = \"\\t\",row.names = FALSE)\n",
        "\n",
        "# Remove everything after the first space, including trailing digits and an optional dot\n",
        "counts.Disc.all <- gsub(\"\\\\s+\\\\d+\\\\.?$\", \"\", counts.Disc.all)\n",
        "\n",
        "# Find elements that will turn into NA when coerced to numeric\n",
        "non_numeric_elements2 <- counts.Disc.all[is.na(as.numeric(counts.Disc.all))]\n",
        "#Warning message:\n",
        "#NAs introduced by coercion\n",
        "\n",
        "length(non_numeric_elements2)\n",
        "#[1] 13\n",
        "\n",
        "print(non_numeric_elements2)\n",
        "#[1] \"11.60540448 7.88\" \"5.455753203 5.36\" \"5.3912438332 5.4\" NA\n",
        "#[5] NA                 NA                 NA                 NA\n",
        "#[9] NA                 NA                 NA                 NA\n",
        "#[13] NA\n",
        "\n",
        "write.table(counts.Disc.all,\"~/NCA.ER/data/pre.curated.Disc.all.Genes.txt\",sep = \"\\t\",row.names = TRUE)\n",
        "\n",
        "#Three non_numeric values the end digits were removed manually\n",
        "Disc.ALL2 <- read.delim(\"~/NCA.ER/Finalyzing/1.Curating.Data/2.pre.counts.Disc.all.Genes.txt\",header = TRUE, as.is = TRUE, row.names = 1)\n",
        "str(Disc.ALL2)\n",
        "#'data.frame':\t19219 obs. of  993 variables:\n",
        "#$ MB_0005: num  5.06 9.92 5.42 6.24 5.23 ...\n",
        "#$ MB_0006: num  5.37 11.31 5.14 6.22 5.91 ...\n",
        "#$ MB_0008: num  5.28 9.83 5.56 6.06 5.3 ...\n",
        "#$ MB_0014: num  5.53 10.68 5.36 6.16 5.4 ...\n",
        "#$ MB_0020: num  5.28 9.31 5.69 6.16 5.53 ...\n",
        "\n",
        "counts.Disc.all <- as.matrix(Disc.ALL2)\n",
        "non_numeric_elements3 <- counts.Disc.all[is.na(as.numeric(counts.Disc.all))]\n",
        "#Warning message:\n",
        "#NAs introduced by coercion\n",
        "\n",
        "length(non_numeric_elements3)\n",
        "#[1] 10\n",
        "\n",
        "print(non_numeric_elements3)\n",
        "#[1] NA NA NA NA NA NA NA NA NA NA\n",
        "\n",
        "# Convert the matrix to numeric values (after applying gsub)\n",
        "counts.Disc.all <- apply(counts.Disc.all, c(1, 2), as.numeric)\n",
        "\n",
        "# Check the structure of the matrix after conversion\n",
        "str(counts.Disc.all)\n",
        "#num [1:19219, 1:993] 5.06 9.92 5.42 6.24 5.23 ...\n",
        "#- attr(*, \"dimnames\")=List of 2\n",
        "#..$ : chr [1:19219] \"A1CF\" \"A2M\" \"A2ML1\" \"A4GALT\" ...\n",
        "#..$ : chr [1:993] \"MB_0005\" \"MB_0006\" \"MB_0008\" \"MB_0014\" ...\n",
        "\n",
        "#Replace NA with the average of the row\n",
        "counts.Disc.all <- t(apply(counts.Disc.all, 1, function(row) {\n",
        "  # Replace NA in the row with the mean of non-NA values in that row\n",
        "  row[is.na(row)] <- mean(row, na.rm = TRUE)\n",
        "  return(row)\n",
        "}))\n",
        "\n",
        "View(counts.Disc.all)\n",
        "\n",
        "write.table(counts.Disc.all,file = file.path(\"~/NCA.ER/Finalyzing/1.Curating.Data/\",\"3.Disc.Genes.Curated.txt\"),sep = \"\\t\")"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#R File from '~/NCA.ER/Finalyzing/1.Curating.Data/Disc.Curating.R'\n",
        "#---Load libraries-----------------\n",
        "library(RTN)\n",
        "#---1.Load Gene Expression Data of 90 metabolic pathways and TFs = a mtrix for counts/assays---------------------------------\n",
        "#load txt. Format for available genes in the Validovery dataset\n",
        "Valid.ALL <- read.delim(\"~/NCA.ER/data/VALID.Genes.txt\",header = TRUE, as.is = TRUE)\n",
        "\n",
        "sum(is.na(Valid.ALL[, 1]))  # Count the number of NA values\n",
        "#[1] 1\n",
        "\n",
        "which(is.na(Valid.ALL[, 1]))  # Identify their positions\n",
        "#[1] 11028 in the main file not sorted, only metabolic clusters row was removed.\n",
        "# [1] 10390 in the sorted data based on gene names without 1 and 0 metabolic clusters.\n",
        "\n",
        "Valid.ALL <- Valid.ALL[-10390,]  # Remove the row with 'NA' as gene name\n",
        "rownames(Valid.ALL) <- Valid.ALL[,1]\n",
        "Valid.ALL <- Valid.ALL[,-1]\n",
        "dim(Valid.ALL)\n",
        "#[1] 19219  994\n",
        "\n",
        "str(Valid.ALL)\n",
        "#'data.frame':\t19219 obs. of  994 variables:\n",
        "#$ MB.0000: num  5.25 9.84 5.34 6.65 5.66 ...\n",
        "#$ MB.0002: num  5.4 8.32 5.36 6.64 5.32 ...\n",
        "#$ MB.0010: num  5.46 8.96 5.55 5.66 5.34 ...\n",
        "#$ MB.0025: num  5.6 10.04 5.18 5.63 5.41 ...\n",
        "#$ MB.0035: num  5.52 8.8 5.05 5.49 5.16 ...\n",
        "\n",
        "View(Valid.ALL)\n",
        "\n",
        "colnames(Valid.ALL) <- gsub(\"\\\\.\", \"_\", colnames(Valid.ALL))\n",
        "\n",
        "#Sort data by columnnames(Sample IDs) and rownames(Gene IDs)\n",
        "Valid.ALL <- Valid.ALL[order(rownames(Valid.ALL)),order(colnames(Valid.ALL))]\n",
        "View(Valid.ALL)\n",
        "head(Valid.ALL)\n",
        "str(Valid.ALL)\n",
        "#'data.frame':\t19219 obs. of  994 variables:\n",
        "#$ MB_0000: num  5.25 9.84 5.34 6.65 5.66 ...\n",
        "#$ MB_0002: num  5.4 8.32 5.36 6.64 5.32 ...\n",
        "#$ MB_0010: num  5.46 8.96 5.55 5.66 5.34 ...\n",
        "#$ MB_0025: num  5.6 10.04 5.18 5.63 5.41 ...\n",
        "#$ MB_0035: num  5.52 8.8 5.05 5.49 5.16 ...\n",
        "\n",
        "#check for NAs\n",
        "sum(is.na(Valid.ALL))\n",
        "#[1] 4\n",
        "\n",
        "which(is.na((rownames(Valid.ALL))))\n",
        "#integer(0)\n",
        "\n",
        "#check for non-numeric elements\n",
        "counts.Valid.all <- as.matrix(Valid.ALL)\n",
        "\n",
        "# Find elements that will turn into NA when coerced to numeric\n",
        "non_numeric_elements1 <- counts.Valid.all[is.na(as.numeric(counts.Valid.all))]\n",
        "\n",
        "length(non_numeric_elements1)\n",
        "#[1] 4\n",
        "\n",
        "print(non_numeric_elements1)\n",
        "#[1] NA NA NA NA\n",
        "\n",
        "# Check the structure of the matrix after conversion\n",
        "str(counts.Valid.all)\n",
        "#num [1:19219, 1:994] 5.25 9.84 5.34 6.65 5.66 ...\n",
        "#- attr(*, \"dimnames\")=List of 2\n",
        "#..$ : chr [1:19219] \"A1CF\" \"A2M\" \"A2ML1\" \"A4GALT\" ...\n",
        "#..$ : chr [1:994] \"MB_0000\" \"MB_0002\" \"MB_0010\" \"MB_0025\" ...\n",
        "\n",
        "counts.Valid.all <- t(apply(counts.Valid.all, 1, function(row) {\n",
        "  # Replace NA in the row with the mean of non-NA values in that row\n",
        "  row[is.na(row)] <- mean(row, na.rm = TRUE)\n",
        "  return(row)\n",
        "}))\n",
        "\n",
        "# Find elements that will turn into NA when coerced to numeric\n",
        "non_numeric_elements2 <- counts.Valid.all[is.na(as.numeric(counts.Valid.all))]\n",
        "\n",
        "length(non_numeric_elements2)\n",
        "#[1] 0\n",
        "\n",
        "print(non_numeric_elements2)\n",
        "#numeric(0)\n",
        "\n",
        "View(counts.Valid.all)\n",
        "\n",
        "write.table(counts.Valid.all,file = file.path(\"~/NCA.ER/Finalyzing/1.Curating Data/Valid.Curating Data/\",\"Valid.Genes.Curated.txt\"),sep = \"\\t\")\n",
        "\n",
        "#---Checking the expression of new combined metabolic genes with all genes---------\n",
        "\n",
        "#Check the data before and after curating nin_numric values\n",
        "Valid.mp.tf <- read.delim(\"~/NCA.ER/data/Valid.MP.TF.txt\",header = TRUE, as.is = TRUE, row.names = 1)\n",
        "colnames(Valid.mp.tf) <- gsub(\"\\\\.\", \"_\", colnames(Valid.mp.tf))\n",
        "Valid.mp.tf <- Valid.mp.tf[order(rownames(Valid.mp.tf)),order(colnames(Valid.mp.tf))]\n",
        "\n",
        "counts.Valid.mp.tf <- as.matrix(Valid.mp.tf)\n",
        "\n",
        "non_numeric_elements.mp.tf <- counts.Valid.mp.tf[is.na(as.numeric(counts.Valid.mp.tf))]\n",
        "print(non_numeric_elements.mp.tf)\n",
        "#numeric(0)\n",
        "\n",
        "#Valid.genes.all-->curated-->Valid.mp.tf OR Valid.genes.all-->Valid.mp.tf-->curated\n",
        "test <- intersect(rownames(counts.Valid.all),rownames(Valid.mp.tf))\n",
        "test2 <- counts.Valid.all[test,]\n",
        "all.equal(counts.Valid.mp.tf,test2) #[1] TRUE\n",
        "\n",
        "save.image(\"~/NCA.ER/Finalyzing/1.Curating Data/Valid.Curating Data/Valid.all.Genes.Curated.RData\")\n"
      ],
      "metadata": {
        "id": "OPlowd6-pXU4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#2.Which Data to Use regarding Discovery Data set?\n",
        "At the first step to compare the expression data, It beginned with loading the disc.mp.tf which stands for discovery data set metabolic genes and transcription factor genes I wanted to show non_numeric elements that have been curated to get the expression matrix that here is named as \"test\" data.\n",
        "\n",
        "After curating the Discovery dataset from non_numeric elements;\n",
        "I checked the difference between the time I took metabolic genes and transcription factor genes directly from discovery data set before curating which I have used through all my analysis here named as \"test\" VS. the time I took the same genes expression after curating discovery data set named as \"test3\". they have 3.667215e-08 mean difference indeed.\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "otBerHwcSBjw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#---Checking the expression of new combined metabolic genes with all genes---------\n",
        "\n",
        "#Check the data before and after curating nin_numric values\n",
        "disc.mp.tf <- read.delim(\"~/NCA.ER/data/DISC.MP.TF.Genes.txt\",header = TRUE, as.is = TRUE, row.names = 1)\n",
        "counts.disc.mp.tf <- as.matrix(disc.mp.tf)\n",
        "\n",
        "non_numeric_elements.mp.tf <- counts.disc.mp.tf[is.na(as.numeric(counts.disc.mp.tf))]\n",
        "print(non_numeric_elements.mp.tf)\n",
        "#[1] \"5.24408565936 5.\" \"5.266270898297 5\" \"7.08498028378 6.\" \"5.475669392427 5\"\n",
        "#[5] \"5.529674338976 5\" \"6.530990146544 7\" \"6.245621615493 6\" \"5.748945951723 5\"\n",
        "#[9] \"6.67129437167 7.\" \"5.883760590443 5\" \"7.607295123545 8\" \"6.391411320326 6\"\n",
        "#[13] \"5.261103082174 5\" \"6.419634552761 5\"\n",
        "\n",
        "#Disc.genes.all-->curated-->disc.mp.tf OR disc.genes.all-->disc.mp.tf-->curated\n",
        "test <- read.delim(\"~/NCA.ER/data/DISC.MP.TF.Curated.txt\",header = TRUE, as.is = TRUE, row.names = 1)\n",
        "colnames(test) <- gsub(\"\\\\.\", \"_\", colnames(test))\n",
        "test <- test[order(rownames(test)),order(colnames(test))]\n",
        "test <- as.matrix(test)\n",
        "test2 <- intersect(rownames(counts.Disc.all),rownames(test))\n",
        "test3 <- counts.Disc.all[test2,]\n",
        "all.equal(test,test3) #[1] \"Mean relative difference: 3.667215e-08\"\n",
        "\n",
        "\n",
        "save.image(\"~/NCA.ER/Finalyzing/1.Curating.Data/Disc.all.Genes.Curated.RData\")\n"
      ],
      "metadata": {
        "id": "Z5Q6oOVCSAZi"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#3.Sample Annotation(Column Annotation)\n",
        "METABRIC Clinical file was retrieved from [cBioPortal](https://www.cbioportal.org/study/clinicalData?id=brca_metabric)\n"
      ],
      "metadata": {
        "id": "T4C7sq_9cKyQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "Metabric_Manual_disc <- read.delim(\"~/NCA/data/METABRIC Clinical.txt\", row.names = 1)\n",
        "View(Metabric_Manual_disc)\n",
        "\n",
        "Metabric_Manual_disc <- Metabric_Manual_disc[order(rownames(Metabric_Manual_disc)),]\n",
        "View(Metabric_Manual_disc)\n",
        "\n",
        "# Replace dots with underscores in column names if needed\n",
        "rownames(Metabric_Manual_disc) <- gsub(\"\\\\-\", \"_\", rownames(Metabric_Manual_disc))\n",
        "View(Metabric_Manual_disc)\n",
        "\n",
        "# Load required library\n",
        "library(dplyr)\n",
        "\n",
        "# Assuming your dataset is named Metabric_Manual_disc\n",
        "# Create a new binary dataframe based on transformations\n",
        "Metabric_Manual_disc <- Metabric_Manual_disc %>%\n",
        "  transmute(\n",
        "    IDs = rownames(Metabric_Manual_disc),\n",
        "    Cohort = Cohort,\n",
        "\n",
        "    OS.time = Overall.Survival..Months.,\n",
        "    OS.event = ifelse(Overall.Survival.Status == \"1:DECEASED\", 1, 0),\n",
        "    DSS.event = ifelse(Patient.s.Vital.Status == \"Died of Disease\", 1, 0),\n",
        "    RFS.event = ifelse(Relapse.Free.Status == \"1:Recurred\", 1, 0),\n",
        "    RFS.time = Relapse.Free.Status..Months.,\n",
        "\n",
        "    Grade = Neoplasm.Histologic.Grade,\n",
        "    Size = Tumor.Size,\n",
        "    LN = Lymph.nodes.examined.positive,\n",
        "    Age = Age.at.Diagnosis,\n",
        "    LN = Lymph.nodes.examined.positive,\n",
        "    Age = Age.at.Diagnosis,\n",
        "\n",
        "\n",
        "    # Subtypes for LumA, LumB, Basal, Her2, Normal based on Pam50 subtype\n",
        "    LumA = ifelse(Pam50...Claudin.low.subtype == \"LumA\", 1, 0),\n",
        "    LumB = ifelse(Pam50...Claudin.low.subtype == \"LumB\", 1, 0),\n",
        "    Basal = ifelse(Pam50...Claudin.low.subtype == \"Basal\", 1, 0),\n",
        "    Her2 = ifelse(Pam50...Claudin.low.subtype == \"Her2\", 1, 0),\n",
        "    Normal = ifelse(Pam50...Claudin.low.subtype == \"Normal\", 1, 0),\n",
        "\n",
        "    # ER and PR status (positive and negative)\n",
        "    `ER+` = ifelse(ER.Status == \"Positive\", 1, 0),\n",
        "    `ER-` = ifelse(ER.Status == \"Negative\", 1, 0),\n",
        "\n",
        "\n",
        "    # Histologic Grade categories G1, G2, G3\n",
        "    G1 = ifelse(Neoplasm.Histologic.Grade == 1, 1, 0),\n",
        "    G2 = ifelse(Neoplasm.Histologic.Grade == 2, 1, 0),\n",
        "    G3 = ifelse(Neoplasm.Histologic.Grade == 3, 1, 0),\n",
        "\n",
        "    # Hormone Therapy (HT)\n",
        "    HT = ifelse(Hormone.Therapy == \"YES\", 1, 0),\n",
        "\n",
        "  )\n",
        "View(Metabric_Manual_disc)\n",
        "\n",
        "# Print head of the colAnnotation for verification\n",
        "head(Metabric_Manual_disc)"
      ],
      "metadata": {
        "id": "VlV4CHijcLjX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#4.Gene Annotation\n",
        "\n",
        "\tensembl_gene_id\tchromosome_name\tstart_position\tend_position\tstrand\texternal_gene_name\tdescription\n",
        "\tENSG00000275841\tHSCHR19_4_CTG3_1\t270964\t275840\t-1\tLILRA3\tleukocyte immunoglobulin like receptor A3 [Source:HGNC Symbol;Acc:HGNC:6604]\n",
        "\tENSG00000288357\tHSCHR19KIR_CA01-TB01_CTG3_1\t134766\t160497\t1\tKIR2DL5A\tkiller cell immunoglobulin like receptor, two Ig domains and long cytoplasmic tail 5A [Source:HGNC Symbol;Acc:HGNC:16345]\n",
        "\tENSG00000288487\tHSCHR19KIR_CA01-TB04_CTG3_1\t134746\t160298\t1\tKIR2DS3\tkiller cell immunoglobulin like receptor, two Ig domains and short cytoplasmic tail 3 [Source:HGNC Symbol;Acc:HGNC:6335]\n",
        "\tENSG00000230463\tHSCHR6_MHC_APD_CTG1\t3824514\t3837642\t-1\tHLA-DRB3\tmajor histocompatibility complex, class II, DR beta 3 [Source:HGNC Symbol;Acc:HGNC:4951]\n",
        "\tENSG00000227826\tHSCHR6_MHC_MANN_CTG1\t3840435\t3855431\t-1\tHLA-DRB4\tmajor histocompatibility complex, class II, DR beta 4 [Source:HGNC Symbol;Acc:HGNC:4952]\n",
        "\n",
        "5 Gene were missed from all genes analysis at the first time \"Since the total number of genes are high and to be able to track reproducibility\" these genes were ignored at the second run as well.\n",
        "\n",
        "This is the suggested code instead of the codes for this section:\n",
        "But according to browseVignettes(\"biomaRt\") [link text](http://127.0.0.1:31622/library/biomaRt/doc/accessing_ensembl.html): I would prefer to use the command with useEnsembl() instead of usemart()\n",
        "\n",
        "> If you are using biomaRt directly make sure you are using useEnsembl() to create the Mart object, rather than useMart(). useEnsembl() is aware of some specific connection details required to connect to Ensembl, and using it may fix any connection problems without requiring you to do anything further.\n",
        "\n",
        "\n",
        "mart <- useMart('ENSEMBL_MART_ENSEMBL')\n",
        "mart <- useDataset('hsapiens_gene_ensembl', mart)\n",
        "\n",
        "annot <- getBM(values = rownames(DISC.Genes), filters = \"external_gene_name\",mart = mart,\n",
        "               attributes = c(\"ensembl_gene_id\", \"chromosome_name\",\n",
        "                              \"start_position\", \"end_position\", \"strand\",\"gene_biotype\",\n",
        "                              \"external_gene_name\",\"description\"))"
      ],
      "metadata": {
        "id": "aiQreEAVnV4B"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#---4.Load Ensemble Gene Annotation----------------------------------------------\n",
        "\n",
        "#1.Accessing the data available in Ensembl by biomaRT\n",
        "browseVignettes(\"biomaRt\")\n",
        "\n",
        "#2.Selecting an Ensembl BioMart database and dataset\n",
        "\n",
        "# Connect to Ensembl database, and query human genes\n",
        "ensembl <- useEnsembl(biomart = \"genes\", dataset = \"hsapiens_gene_ensembl\")\n",
        "#Ensembl site unresponsive, trying asia mirror\n",
        "\n",
        "\n",
        "#2.1##Step1: Identifying the database you need\n",
        "\n",
        "#find the names of the BioMart services Ensembl is currently providing\n",
        "##and Can be used to connect to the desired BioMart database:\n",
        "listEnsembl()\n",
        "#biomart                version\n",
        "#1         genes      Ensembl Genes 113\n",
        "#2 mouse_strains      Mouse strains 113\n",
        "#3          snps  Ensembl Variation 113\n",
        "#4    regulation Ensembl Regulation 113\n",
        "\n",
        "#biomart argument should be given a valid name from the output of listEnsembl()\n",
        "ensembl <- useEnsembl(biomart = \"genes\")\n",
        "#Ensembl site unresponsive, trying useast mirror\n",
        "\n",
        "#2.2##Step 2: Choosing a dataset\n",
        "\n",
        "#Within the Ensembl dataset each species is a different dataset.\n",
        "#look at which datasets are available in the selected BioMart\n",
        "#by using the function listDatasets()\n",
        "datasets <- listDatasets(ensembl)\n",
        "View(datasets)\n",
        "\n",
        "##Because the listDatasets are so long\n",
        "##in ensembl mart find anything that contain the term ‘hsapiens’\n",
        "\n",
        "searchDatasets(mart = ensembl, pattern = \"hsapiens\")\n",
        "#dataset              description    version\n",
        "#80 hsapiens_gene_ensembl Human genes (GRCh38.p14) GRCh38.p14\n",
        "\n",
        "#If you've been through these all before;select a both the database and dataset in one step\n",
        "ensembl <- useDataset(dataset = \"hsapiens_gene_ensembl\", mart = ensembl)\n",
        "\n",
        "\n",
        "#2.3##Ensembl mirror sites\n",
        "#listMarts() to find the biomart\n",
        "ensembl <- useEnsembl(biomart = \"ENSEMBL_MART_ENSEMBL\",\n",
        "                      dataset = \"hsapiens_gene_ensembl\",\n",
        "                      mirror = \"useast\")\n",
        "#Ensembl site unresponsive, trying asia mirror\n",
        "#maintain consistent annotation throughout the duration of a project.\n",
        "\n",
        "#2.4##Using archived versions of Ensembl, with no arguments\n",
        "listEnsemblArchives()\n",
        "#name     date                                 url version current_release\n",
        "#1  Ensembl GRCh37 Feb 2014          https://grch37.ensembl.org  GRCh37\n",
        "#2     Ensembl 113 Oct 2024 https://oct2024.archive.ensembl.org     113               *\n",
        "#  3     Ensembl 112 May 2024 https://may2024.archive.ensembl.org     112\n",
        "#4     Ensembl 111 Jan 2024 https://jan2024.archive.ensembl.org     111\n",
        "#5     Ensembl 110 Jul 2023 https://jul2023.archive.ensembl.org     110\n",
        "#6     Ensembl 109 Feb 2023 https://feb2023.archive.ensembl.org     109\n",
        "#7     Ensembl 108 Oct 2022 https://oct2022.archive.ensembl.org     108\n",
        "#8     Ensembl 107 Jul 2022 https://jul2022.archive.ensembl.org     107\n",
        "#9     Ensembl 106 Apr 2022 https://apr2022.archive.ensembl.org     106\n",
        "#10    Ensembl 105 Dec 2021 https://dec2021.archive.ensembl.org     105\n",
        "#11    Ensembl 104 May 2021 https://may2021.archive.ensembl.org     104\n",
        "#12    Ensembl 103 Feb 2021 https://feb2021.archive.ensembl.org     103\n",
        "#13    Ensembl 102 Nov 2020 https://nov2020.archive.ensembl.org     102\n",
        "#14    Ensembl 101 Aug 2020 https://aug2020.archive.ensembl.org     101\n",
        "#15    Ensembl 100 Apr 2020 https://apr2020.archive.ensembl.org     100\n",
        "#16     Ensembl 99 Jan 2020 https://jan2020.archive.ensembl.org      99\n",
        "#17     Ensembl 98 Sep 2019 https://sep2019.archive.ensembl.org      98\n",
        "#19     Ensembl 77 Oct 2014 https://oct2014.archive.ensembl.org      77\n",
        "#20     Ensembl 75 Feb 2014 https://feb2014.archive.ensembl.org      75\n",
        "#21     Ensembl 54 May 2009 https://may2009.archive.ensembl.org      54\n",
        "\n",
        "listEnsembl(version = 113)\n",
        "#biomart                version\n",
        "#1         genes      Ensembl Genes 113\n",
        "#2 mouse_strains      Mouse strains 113\n",
        "#3          snps  Ensembl Variation 113\n",
        "#4    regulation Ensembl Regulation 113\n",
        "\n",
        "#to ensure that script you write now will return exactly the same results in the future;\n",
        "##copy the URL from listEnsemblArchive() or www.ensembl.org in host Argument\n",
        "ensembl_113 <- useEnsembl(biomart = \"genes\",\n",
        "                          dataset = \"hsapiens_gene_ensembl\",\n",
        "                          host = \"https://oct2024.archive.ensembl.org\",\n",
        "                          version = 113)\n",
        "\n",
        "searchDatasets(mart = ensembl_113, pattern = \"hsapiens\")\n",
        "#dataset              description    version\n",
        "#80 hsapiens_gene_ensembl Human genes (GRCh38.p14) GRCh38.p14\n",
        "\n",
        "# Extract gene annotations for the corresponding genes from `ensembl_113`\n",
        "filters = listFilters(ensembl_113)\n",
        "View(filters)\n",
        "\n",
        "View(listAttributes(ensembl_113))\n",
        "\n",
        "gene_annot_all_Disc <- getBM(\n",
        "  attributes = c(\"ensembl_gene_id\", \"chromosome_name\", \"start_position\", \"end_position\", \"strand\", \"external_gene_name\",\"description\",\"gene_biotype\"),\n",
        "  filters = \"external_gene_name\",\n",
        "  values = rownames(counts.Disc.all),\n",
        "  mart = ensembl_113\n",
        ")\n",
        "View(gene_annot_all_Disc)\n",
        "dim(gene_annot_all_Disc) #[1] 21574     8\n",
        "\n",
        "#WITH ENTREZ 3185 AND WITHOUT THAT 3152, DON'T MENTION \"entrezgene_id\".\n",
        "\n",
        "# Check for duplicated external_gene_names\n",
        "table(duplicated(gene_annot_all_Disc$external_gene_name))\n",
        "#FALSE  TRUE\n",
        "#18520  3054\n",
        "\n",
        "#To check with Previously retrieved gene annotations\n",
        "\n",
        "#annotation <- read.delim(\"~/NCA.ER/Gene_annot_all.metabric.biotype.retrieved.txt\",header = TRUE, as.is = TRUE)\n",
        "#identical(annotation,gene_annot_all_Disc) #[1] TRUE\n",
        "\n",
        "#annotation2 <- read.delim(\"~/NCA.ER/Gene_annot_all.metabric.retrieved.txt\",header = TRUE, as.is = TRUE)\n",
        "#identical(gene_annot_all_Disc[,-8],annotation2) #[1] TRUE\n",
        "\n",
        "write.table(gene_annot_all_Disc1,file = file.path(\"~/NCA.ER/\",\"Gene_annot_all.Disc.biotype.retrieved.txt\"),sep = \"\\t\",row.names = FALSE)\n",
        "save.image(\"~/NCA.ER/NCA.Disc/NCA.Disc/up.to.gene.annot.RData\")\n",
        "\n",
        "#---Find common IDs from gene annotation and data set---\n",
        "# Extract gene IDs from both datasets\n",
        "ensemble_ids <- gene_annot_mp.tf_disc$external_gene_name\n",
        "View(ensemble_ids)\n",
        "\n",
        "ensemble_ids <- sort(ensemble_ids)\n",
        "length(ensemble_ids)\n",
        "\n",
        "disc_mp.tf_gene_ids <- rownames(counts.mp.tf.disc)\n",
        "View(disc_mp.tf_gene_ids)\n",
        "\n",
        "common_genes <- intersect(ensemble_ids, disc_mp.tf_gene_ids) # Should match or be close to 1420\n",
        "common_genes <- sort(common_genes)\n",
        "all.equal(ensemble_ids, common_genes)\n",
        "length(common_genes)\n",
        "\n",
        "# Subset and reorder both datasets to only include common samples\n",
        "counts.mp.tf.disc <- counts.mp.tf.disc[common_genes, ]\n",
        "View(counts.mp.tf.disc)\n",
        "\n",
        "all.equal(ensemble_ids,rownames(counts.mp.tf.disc))\n",
        "\n",
        "#Set rownames for gene annotation, %n% doesn't let your data go NA\n",
        "gene_annot_mp.tf_disc <- gene_annot_mp.tf_disc[gene_annot_mp.tf_disc$external_gene_name %in% common_genes, ]\n",
        "rownames(gene_annot_mp.tf_disc) <- gene_annot_mp.tf_disc$external_gene_name\n",
        "View(gene_annot_mp.tf_disc)\n",
        "all.equal(gene_annot_mp.tf_disc$external_gene_name, rownames(counts.mp.tf.disc))\n",
        "all.equal(gene_annot_mp.tf_disc$external_gene_name,rownames(gene_annot_mp.tf_disc))\n",
        "all.equal(gene_annot_mp.tf_disc$external_gene_name,common_genes)\n",
        "all.equal(common_genes,ensemble_ids)\n",
        "\n",
        "#Sort counts.mp.tf.disc and gene_annot_mp.tf_disc based on rownames which is gene names\n",
        "counts.mp.tf.disc <- counts.mp.tf.disc[order(rownames(counts.mp.tf.disc)), ]\n",
        "gene_annot_mp.tf_disc <- gene_annot_mp.tf_disc[order(rownames(gene_annot_mp.tf_disc)), ]\n",
        "\n",
        "all.equal(rownames(counts.mp.tf.disc), rownames(gene_annot_mp.tf_disc))  # Should return TRUE if aligned\n",
        "dim(counts.mp.tf.disc)\n",
        "dim(gene_annot_mp.tf_disc)\n",
        "View(gene_annot_mp.tf_disc)\n",
        "\n",
        "#In rtni analysis it needs \"SYMBOL\" in rowAnnotation\n",
        "colnames(gene_annot_mp.tf_disc)\n",
        "colnames(gene_annot_mp.tf_disc)[colnames(gene_annot_mp.tf_disc) == \"external_gene_name\"] <- \"SYMBOL\"\n",
        "colnames(gene_annot_mp.tf_disc)[colnames(gene_annot_mp.tf_disc) == \"ensembl_gene_id\"] <- \"ENSEMBL\"\n",
        "\n",
        "# Verify the change\n",
        "colnames(gene_annot_mp.tf_disc)\n",
        "\n",
        "\n",
        "# One final check:\n",
        "stopifnot(rownames(gene_annot_mp.tf_disc) == rownames(counts.mp.tf.disc), # features\n",
        "          rownames(sampleAnnotation.disc) == colnames(counts.mp.tf.disc)) # samples\n"
      ],
      "metadata": {
        "id": "DLoLXqIRbkyU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#5.Build First rtni-data"
      ],
      "metadata": {
        "id": "_HmzcQkbeXY3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#---Build a list of expression data along with row and column annotation---\n",
        "Disc_all <- list(\n",
        "  expData = counts.Disc.all,                       # expData as an assay\n",
        "  rowAnnotation = Disc.Gene.Annot.Curated,             # Gene annotations (row metadata)\n",
        "  colAnnotation = sampleAnnotation.Disc           # Sample annotations (column metadata)\n",
        ")\n",
        "\n",
        "View(Disc_all)\n",
        "\n",
        "#---Load RegulatoryElements---\n",
        "library(RTN)\n",
        "# Load TF annotation\n",
        "data(\"tfsData\")\n",
        "\n",
        "# Check TF annotation:\n",
        "# Intersect TFs from Lambert et al. (2018) with gene annotation\n",
        "# from the gene expression of 90 metabolic pathway cohort\n",
        "regulatoryElements <- intersect(tfsData$Lambert2018$SYMBOL, Disc_all$rowAnnotation$SYMBOL)\n",
        "View(regulatoryElements)\n",
        "regulatoryElements <- sort(regulatoryElements)\n",
        "View(regulatoryElements)\n",
        "\n",
        "#---constructing rtni---\n",
        "#This dataset consists of a list with 3 objects:\n",
        "##a named gene expression matrix (tniData$expData),\n",
        "##a data frame with gene annotations (tniData$rowAnnotation),\n",
        "##and a data frame with sample annotations (tniData$colAnnotation).\n",
        "##alternatively, 'expData' can be a 'SummarizedExperiment' object\n",
        "rtni_disc_mp.tf <- tni.constructor(expData = disc_mp.tf$expData,\n",
        "                             regulatoryElements = regulatoryElements,\n",
        "                             rowAnnotation = disc_mp.tf$rowAnnotation,\n",
        "                             colAnnotation = disc_mp.tf$colAnnotation)\n",
        "\n",
        "save(rtni_disc_mp.tf, file = file.path(\"~/NCA.ER/NCA.MP.TF.DISC.NCA/\", \"rtni_mp.tf.disc.RData\"))\n"
      ],
      "metadata": {
        "id": "wbZK3uVSf3qt"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#6.Building Regulons\n"
      ],
      "metadata": {
        "id": "xDCQnICdrkle"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "###snow clustering returns zero results\n",
        "### Compute the reference regulatory network by permutation and bootstrap analyses.\n",
        "### Please set 'spec' according to your available hardware\n",
        "#options(cluster=snow::makeCluster(spec=4, \"SOCK\"))  #???\n",
        "\n",
        "#1.with pValuCutoff =1e-7 it gives 0 values, even in the summary\n",
        "#rtni_Disc_all.NA <- tni.permutation(rtni_Disc_all.NA, pValueCutoff = 1e-7) #pValueCutoff=1e-7 ?? zero_values\n",
        "\n",
        "#2.with pValuCutoff= 1e-7 and npermutation = 1000 it gives 0 values, even in the sumary\n",
        "#rtni_Disc_all.NA <- tni.permutation(rtni_Disc_all.NA,nPermutations = 1000, pValueCutoff = 1e-7) #pValueCutoff=1e-7 ?? zero_values\n",
        "\n",
        "#3.  nPermutations >= 1000 with running snow goes the same zero values\n",
        "#4.  nPermutations >= 1000 without snow package, and pValueCutoff works!\n",
        "rtni_Disc_all <- tni.permutation(rtni_Disc_all, nPermutations = 1000)\n",
        "\n",
        "#Unstable interactions are subsequently removed by bootstrap analysis,\n",
        "##creates a consensus bootstrap network, referred here as refnet (reference network).\n",
        "rtni_Disc_all <- tni.bootstrap(rtni_Disc_all)\n",
        "\n",
        "#stopCluster(getOption(\"cluster\"))  # avoid leak memory\n",
        "###############################################################\n",
        "\n",
        "# Compute the DPI-filtered regulatory network\n",
        "rtni_Disc_all.NA <- tni.dpi.filter(rtni_Disc_all, eps = NA)\n",
        "tni.regulon.summary(rtni_Disc_all.NA)\n",
        "#Regulatory network comprised of 1508 regulons.\n",
        "#-- DPI-filtered network:\n",
        "#  regulatoryElements            Targets              Edges\n",
        "#1508              17367             135358\n",
        "#Min. 1st Qu.  Median    Mean 3rd Qu.    Max.\n",
        "#0.0    34.0    69.0    89.8   120.0   811.0\n",
        "#-- Reference network:\n",
        "#  regulatoryElements            Targets              Edges\n",
        "#1508              17367            4225619\n",
        "#Min. 1st Qu.  Median    Mean 3rd Qu.    Max.\n",
        "#0     787    2624    2802    4568    7974\n",
        "#---\n",
        "save.image(\"~/NCA.ER/NCA.Disc.All.Genes/Disc.epsNA.RData\")\n",
        "\n",
        "#detailed information about a specific regulon\n",
        "tni.regulon.summary(rtni_Disc_all.NA, regulatoryElements = \"GATA3\")\n",
        "#The GATA3 regulon has 352 targets, it's a large and unbalanced regulon.\n",
        "#-- DPI filtered network targets:\n",
        "#  Total Positive Negative\n",
        "#352       44      308\n",
        "#-- Reference network targets:\n",
        "#  Total Positive Negative\n",
        "#7426     3527     3899\n",
        "#-- Regulators with mutual information:\n",
        "#  AKAP8, AKAP8L, AKNA, AR, ARID2, ARID3B, ARID5B, ARNT2, ASCL1, ASCL2...[537 more]\n",
        "#---\n",
        "# Save the TNI object for subsequent analyses\n",
        "save(rtni_Disc_all.NA, file = file.path(\"~/NCA.ER/NCA.all.Disc.NCA/epsData/\", \"rtni_Disc_all.NA.epsNA.RData\"))\n",
        "\n",
        "print(\"Step 8 completed: Loaded rtni data and saved.\")\n",
        "\n",
        "regulon.NA <- tni.get(rtni_VALID_all.NA, what = 'regulons.and.mode', idkey = \"SYMBOL\")\n",
        "View(regulon.NA)\n",
        "head(regulon.NA)\n",
        "\n",
        "#---Find the maximum number of genes across all regulons to extract the regulon in the txt. format ---\n",
        "max_genes <- max(sapply(regulon.NA, function(x) if (is.null(x)) 0 else length(x)))\n",
        "\n",
        "# Create a list of data frames with equal row lengths\n",
        "regulon_list <- lapply(names(regulon.NA), function(regulon_name) {\n",
        "  regulon_data <- regulon.NA[[regulon_name]]\n",
        "\n",
        "  # Handle empty or NULL regulons\n",
        "  if (is.null(regulon_data) || length(regulon_data) == 0) {\n",
        "    df <- data.frame(\n",
        "      Gene = rep(NA, max_genes),\n",
        "      Value = rep(NA, max_genes),\n",
        "      stringsAsFactors = FALSE\n",
        "    )\n",
        "  } else if (is.vector(regulon_data)) {\n",
        "    # Ensure matching lengths for Gene and Value\n",
        "    genes <- names(regulon_data)\n",
        "    values <- regulon_data\n",
        "    if (length(genes) == 0) genes <- rep(NA, length(values))\n",
        "    df <- data.frame(\n",
        "      Gene = genes,\n",
        "      Value = values,\n",
        "      stringsAsFactors = FALSE\n",
        "    )\n",
        "  } else if (is.matrix(regulon_data) || is.data.frame(regulon_data)) {\n",
        "    df <- data.frame(\n",
        "      Gene = rownames(regulon_data),\n",
        "      Value = regulon_data[, 1], # Assuming values are in the first column\n",
        "      stringsAsFactors = FALSE\n",
        "    )\n",
        "  } else {\n",
        "    df <- data.frame(\n",
        "      Gene = rep(NA, max_genes),\n",
        "      Value = rep(NA, max_genes),\n",
        "      stringsAsFactors = FALSE\n",
        "    )\n",
        "  }\n",
        "\n",
        "  # Extend to max_genes rows if needed\n",
        "  if (nrow(df) < max_genes) {\n",
        "    df <- rbind(df, data.frame(\n",
        "      Gene = rep(NA, max_genes - nrow(df)),\n",
        "      Value = rep(NA, max_genes - nrow(df))\n",
        "    ))\n",
        "  }\n",
        "\n",
        "  # Rename columns with regulon names\n",
        "  colnames(df) <- c(paste0(regulon_name, \"_Gene\"), paste0(regulon_name, \"_Value\"))\n",
        "  return(df)\n",
        "})\n",
        "\n",
        "# Combine all into one data frame\n",
        "regulon_df <- do.call(cbind, regulon_list)\n",
        "\n",
        "# Write to file\n",
        "write.table(regulon_df, file = (\"Valid.regulon.NA.txt\"), sep = \"\\t\", row.names = FALSE, quote = FALSE)\n",
        "save(regulon.NA,file = \"Valid.All.Genes.Regulon.NA.RData\")\n",
        "\n",
        "rtni_Disc_all.NA_all <- tni.gsea2(rtni_disc_mp.tf, regulatoryElements = rtni_disc_mp.tf@regulatoryElements)\n",
        "disc_regact_disc <- tni.get(rtni_disc_mp.tf, what = \"regulonActivity\")\n",
        "View(disc_regact_disc)"
      ],
      "metadata": {
        "id": "EvIdVFnUrlK_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "0jAndSMO0S6C"
      }
    }
  ]
}